File: 2020/day01-1.ts
function solution(content: string) {
  return find(
    content
      .split("\n")
      .map((item) => parseInt(item))
      .toSorted()
  );
}

function find(list: number[]) {
  const fst = list.at(0)!;
  const lst = list.at(-1)!;
  const result = fst + lst;
  if (result > 2020) return find(list.slice(0, -1));
  if (result < 2020) return find(list.slice(1));
  return fst * lst;
}

Deno.readTextFile("day01.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2020/day01_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return find(sorted(int(x) for x in content if x))


def find(expenses: list[int]):
    fst, *rest, lst = expenses
    result = fst + lst
    if result > 2020:
        return find([fst] + rest)
    if result < 2020:
        return find(rest + [lst])
    return fst * lst


def main() -> None:
    with open("day01.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
